File: app2/py/service.py
import sys
import uuid
from PyQt5.QtCore import QDateTime
from PyQt5.QtNetwork import QTcpSocket
from PyQt5.QtWidgets import QApplication
from reapython.storage import fsStorage
from reapython.rea import scopeCache, stream, pipelines, pipeline, pipeParallel
from reapython.reaRemote import connectRemote, pipelineRemote
from reapython.server import normalServer
from time import sleep
import threading
import json

def readStorage(aRoot: str, aPath: str, aConfig: dict, aType: str) -> stream:
    return pipelines().input(False, "", scopeCache({"path": aPath, "config": aConfig}), True).asyncCall(aRoot + "read" + aType)

def writeStorage(aRoot: str, aPath: str, aConfig: dict, aType: str, aData) -> stream:
    return pipelines().input(False, "", scopeCache({"path": aPath, "config": aConfig, "data": aData}), True).asyncCall(aRoot + "write" + aType)

class dl:
    def __init__(self):
        self.__stgPath = "F:/3M/default/jobs.json"
        self.__stgConfig = {}
        self.__stgRoot = ""
        self.__whole_time = 1000
        self.__jobs = {}
        self.__running = set()
        self.__clients = dict()
        self.__job_mutex = threading.Lock()
        self.__socket_mutex = threading.Lock()

        self.serviceInternal()
        stm = readStorage(self.__stgRoot, self.__stgPath, self.__stgConfig, "JsonObject")
        if stm.data():
            self.__jobs = stm.scope().data("data")
            for i, j in self.__jobs.items():
                if j["status"] == "running":
                    if not self.startJob(i):
                        break
        
        def clientOnline(aInput: stream):
            with self.__socket_mutex:
                self.__clients[aInput.data()] = aInput.scope().data("socket")
                aInput.scope().cache("config", self.__stgConfig).cache("root", self.__stgRoot).cache("path", self.__stgPath)
            aInput.out()
        pipelines("server").add(clientOnline, {"name": "clientOnline", "external": "c++"})

        def clientOffline(aInput: stream):
            if not aInput.data():
                with self.__socket_mutex:
                    for i, j in self.__clients.items():
                        if aInput.scope().data("socket") == j:
                            del self.__clients[i]
                            break
        pipelines().find("clientStatusChanged").nextF(clientOffline)

        def startTrain(aInput: stream):
            usr = self.getUser(aInput.scope().data("socket"))
            dt = aInput.data()
            if usr == "":
                dt["err"] = 1
                dt["msg"] = "no this user!"
                aInput.setData(dt).out()
                return
            job = self.snowflakeID()
            pth = dt["path"]
            root = dt["root"]
            cfg = dt["config"]
            with self.__job_mutex:
                self.__jobs[job] = {"path": pth,
                                    "root": root,
                                    "config": cfg,
                                    "user": usr,
                                    "status": "running"}
                if writeStorage(self.__stgRoot, self.__stgPath, self.__stgConfig, "JsonObject", self.__jobs) and self.startJob(job):
                    dt["err"] = 0
                else:
                    del self.__jobs[job]
                    dt["err"] = 1
                    dt["msg"] = "start job failed!"
                aInput.setData(dt).out()
        pipelines("server").add(startTrain, {"name": "startTrain", "external": "qml"})

    def serviceInternal(self):        
        def execScript(aInput: stream):
            exec(aInput.data())
            aInput.out()
        pipelines("server").add(execScript, {"name": "evalPyScript",
            "external": "c++"})

        def doTraining(aInput: stream):
            tm = aInput.data()
            if tm > self.__whole_time:
                job = aInput.scope().data("job")
                with self.__job_mutex:
                    cfg = self.__jobs[job]
                    cfg["status"] = "success"
                    self.__jobs[job] = cfg
                    while not writeStorage(self.__stgRoot, self.__stgPath, self.__stgConfig, "JsonObject", self.__jobs).data():
                        sleep(0.005)
                    self.__running.remove(job)
            else:
                sleep(0.5)
                tm += 1
                aInput.outs(tm, "doTraining")
            usr = aInput.scope().data("user")
            with self.__socket_mutex:
                if usr in self.__clients:
                    pipelines("server").run("logTrain", 
                                            "time cost: " + str(tm) if tm < self.__whole_time else "complete", 
                                            "", 
                                            aInput.scope().cache("socket", self.__clients[usr]))
        pipelines().add(doTraining, {"name": "doTraining", "type": "Parallel"})
    
    def startJob(self, aJob: str) -> False:
        if len(self.__running) > 4:
            return False
        self.__running.add(aJob)
        cfg = self.__jobs[aJob].copy()
        cfg["job"] = aJob
        cfg["root0"] = self.__stgRoot
        cfg["path0"] = self.__stgPath
        cfg["config0"] = self.__stgConfig
        cfg["remote"] = True
        pipelines().run("doTraining", cfg.get("time", 0), "", scopeCache(cfg))
        return True
    
    def snowflakeID(self):
        return str(QDateTime().currentDateTime().toTime_t()) + "-" + str(uuid.uuid4())

    def getUser(self, aSocket: QTcpSocket)-> str:
        for i, j in self.__clients.items():
            if j == aSocket:
                return i
        return ""

def initServer():
    pipelines().add(lambda aInput:
        aInput.setData(pipeline("server")).out()
    , {"name": "createserverpipeline"})

    pipelines().add(lambda aInput:
        aInput.setData(pipelineRemote("c++", "server")).out()
    , {"name": "createc++pipeline"})

    pipelines().add(lambda aInput:
        aInput.setData(pipelineRemote("qml", "server")).out()
    , {"name": "createqmlpipeline"})

    server = normalServer()
    def writeRemote(aInput: stream):
        server.writeSocket(aInput.scope().data("socket"), aInput.data())
    connectRemote("server", "c++", writeRemote, False, "c++_server")
    connectRemote("server", "qml", writeRemote, False, "qml_server")

def main():
    app = QApplication(sys.argv)
    
    #init storage
    stg = fsStorage()
    stg.initialize()
    #init server sidecar
    initServer()
    #init control
    sv = dl()

    #test.stgTest()
    #test.doTest(1)

    # w = QWidget()
    # w.resize(250, 150)
    # w.move(300, 300)
    # w.setWindowTitle('Simple')
    # w.show()

    sys.exit(app.exec_())

if __name__ == '__main__':
    main()

File: app2/service.cpp
#include "rea.h"
#include "util.h"
#include "reaRemote.h"
#include "storage.h"
#include <QQmlApplicationEngine>
#include <QTcpSocket>
#include <QDateTime>

class dl{
private:
    const QString stgPath = "F:/3M/default/jobs.json";
    const QJsonObject stgConfig = QJsonObject();
    const QString stgRoot = "";
    std::mutex m_job_mutex, m_socket_mutex;
private:
    bool startJob(const QString& aJob){
        if (m_running.size() > 4)
            return false;
        m_running.insert(aJob);
        auto cfg = m_jobs.value(aJob).toObject();
        rea2::pipeline::instance()->run<int>("doTraining", cfg.value("time").toInt(),
                                            "", std::make_shared<rea2::scopeCache>(rea2::Json(cfg,
                                                                                            "job", aJob,
                                                                                            "root0", stgRoot,
                                                                                            "path0", stgPath,
                                                                                            "config0", stgConfig,
                                                                                            "remote", true)));
        return true;
    }
private:
    void serviceTest(const QString& aExternal){
        /*rea2::pipeline::instance()->add<QJsonObject>([](rea2::stream<QJsonObject>* aInput){
            std::cout << "from client" << std::endl;
            aInput->setData(rea2::Json("hello", "world"))->out();
        }, rea2::Json("name", "testServer",
                     "external", aExternal));*/
    }
    const int whole_time = 1000;
    void serviceInternal(){
        /*rea2::pipeline::instance()->add<QString>([](rea2::stream<QString>* aInput){
            aInput->out();
        }, rea2::Json("name", "logTrain",
                     "external", "qml"));*/

        rea2::pipeline::instance()->add<int, rea2::pipeParallel>([this](rea2::stream<int>* aInput){
            auto tm = aInput->data();
            if (tm >= whole_time){
                auto job = aInput->scope()->data<QString>("job");
                std::lock_guard<std::mutex> lg(m_job_mutex);
                auto cfg = m_jobs.value(job).toObject();
                cfg.insert("status", "success");
                m_jobs.insert(job, cfg);
                while (!writeJsonObject(stgRoot, stgPath, stgConfig, m_jobs)->data())
                    std::this_thread::sleep_for(std::chrono::milliseconds(500));
                m_running.remove(job);
            }else{
                std::this_thread::sleep_for(std::chrono::milliseconds(500));
                tm += 1;
                aInput->outs(tm, "doTraining");
            }
            auto usr = aInput->scope()->data<QString>("user");
            std::lock_guard<std::mutex> lg(m_socket_mutex);
            if (m_clients.contains(usr)){
                rea2::pipeline::instance()->run("logTrain", tm < whole_time ? "time cost: " + QString::number(tm) : "complete",
                                               "", aInput->scope()->cache("socket", m_clients.value(usr)));
            }
        }, rea2::Json("name", "doTraining"));
    }
public:
    dl(){
        serviceTest("c++");
        serviceTest("qml");
        serviceInternal();

        auto stm = readStorage(stgRoot, stgPath, stgConfig, "JsonObject");
        if (stm->data()){
            m_jobs = stm->scope()->data<QJsonObject>("data");
            for (auto i : m_jobs.keys()){
                auto job = m_jobs.value(i).toObject();
                if (job.value("status") == "running")
                    if (!startJob(i))
                        break;
            }
        }

        rea2::pipeline::instance()->add<QString>([this](rea2::stream<QString>* aInput){
            std::lock_guard<std::mutex> lg(m_socket_mutex);
            m_clients.insert(aInput->data(), aInput->scope()->data<QTcpSocket*>("socket"));
            aInput->scope()->cache("config", stgConfig)
                    ->cache("root", stgRoot)
                    ->cache("path", stgPath);
            aInput->out();
        }, rea2::Json("name", "clientOnline",
                     "external", "c++"));

        rea2::pipeline::instance()->find("clientStatusChanged")->nextF<bool>([this](rea2::stream<bool>* aInput){
            if (!aInput->data()){
                std::lock_guard<std::mutex> lg(m_socket_mutex);
                for (auto i : m_clients.keys())
                    if (aInput->scope()->data<QTcpSocket*>("socket") == m_clients.value(i)){
                        m_clients.remove(i);
                        break;
                    }
            }
        });

        rea2::pipeline::instance()->add<QJsonObject>([this](rea2::stream<QJsonObject>* aInput){
            auto usr = getUser(aInput->scope()->data<QTcpSocket*>("socket"));
            if (usr == ""){
                aInput->setData(rea2::Json(aInput->data(),
                                          "err", 1,
                                          "msg", "no this user!"))->out();
                return;
            }
            auto dt = aInput->data();
            auto job = snowflakeID();
            auto pth = dt.value("path").toString();
            auto root = dt.value("root").toString();
            auto cfg = dt.value("config").toObject();
            std::lock_guard<std::mutex> lg(m_job_mutex);
            m_jobs.insert(job, rea2::Json("path", pth,
                                         "root", root,
                                         "config", cfg,
                                         "user", usr,
                                         "status", "running"));
            if (writeJsonObject(stgRoot, stgPath, stgConfig, m_jobs)->data() && startJob(job)){
                aInput->setData(rea2::Json(aInput->data(), "err", 0))->out();
            }else{
                m_jobs.remove(job);
                aInput->setData(rea2::Json(aInput->data(),
                                          "err", 1,
                                          "msg", "start job failed!"))->out();
            }
        }, rea2::Json("name", "startTrain", "external", "qml"));
    }
private:
    QString snowflakeID(){
        return QString::number(QDateTime().currentDateTime().toTime_t()) + "-" + rea2::generateUUID();
    }
    QString getUser(QTcpSocket* aSocket){
        for (auto i : m_clients.keys())
            if (m_clients.value(i) == aSocket)
                return i;
        return "";
    }
    QHash<QString, QTcpSocket*> m_clients;
    QSet<QString> m_running;
    QJsonObject m_jobs;
};

static rea2::regPip<QQmlApplicationEngine*> reg_tcp_linker([](rea2::stream<QQmlApplicationEngine*>* aInput){
    static rea2::fsStorage stg;
    stg.initialize();
    static dl dl_;
    aInput->out();
}, rea2::Json("name", "install10_server"), "initRea");
